qcs/src/qpu/execution.rs

```diff
@@ -36,13 +36,105 @@ pub(crate) struct Execution<'a> {
 
 #[derive(Debug, thiserror::Error)]
 pub(crate) enum Error {
-    #[error("problem communicating with QCS")]
-    Qcs {
-        source: Report,
-        retry_after: Option<Duration>,
+    #[error("QPU not found")]
+    QpuNotFound,
+    #[error("QPU currently unavailable, retry after {} seconds", .0.as_secs())]
+    QpuUnavailable(Duration),
+    #[error("Received unauthorized response from QCS, try refreshing credentials")]
+    Unauthorized,
+    #[error("Error communicating with QCS, check network connection and QCS status")]
+    QcsCommunication,
+    #[error("QCS returned an unrecognized error: {0}")]
+    Qcs(String),
+    #[error("problem processing the provided Quil: {0}")]
+    Quil(String),
+    #[error("An error likely caused by a bug in this library or QCS")]
+    Bug(#[from] Bug),
+    #[error("Problem communicating with quilc at {uri}: {details}")]
+    Quilc { uri: String, details: String },
+}
+
+impl From<quilc::Error> for Error {
+    fn from(source: quilc::Error) -> Self {
+        match source {
+            quilc::Error::Isa(source) => Self::Bug(Bug::Isa(format!("{:?}", source))),
+            quilc::Error::QuilcConnection(uri, details) => Self::Quilc {
+                uri,
+                details: format!("{:?}", details),
+            },
+            quilc::Error::QuilcCompilation(details) => Self::Quil(details),
+        }
+    }
+}
+
+impl From<engagement::Error> for Error {
+    fn from(source: engagement::Error) -> Self {
+        match source {
+            engagement::Error::QuantumProcessorUnavailable(duration) => {
+                Self::QpuUnavailable(duration)
+            }
+            engagement::Error::Unauthorized => Self::Unauthorized,
+            engagement::Error::Connection(_) => Self::QcsCommunication,
+            engagement::Error::Schema(_)
+            | engagement::Error::Unknown(_)
+            | engagement::Error::Internal(_) => Self::Bug(Bug::Qcs(format!("{:?}", source))),
+        }
+    }
+}
+
+impl From<runner::Error> for Error {
+    fn from(source: runner::Error) -> Self {
+        match source {
+            runner::Error::Connection(_) => Self::QcsCommunication,
+            runner::Error::Bug(err) => Self::Bug(Bug::Qcs(format!("{:?}", err))),
+            runner::Error::Qpu(err) => Self::Qcs(err),
+        }
+    }
+}
+
+/// Errors that are likely due to bugs in this library or QCS.
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum Bug {
```

Comment:
nit: maybe classify these as `Internal` or something. Although Bug does make sense, I feel like an end-user might ask "if you knew this was a bug, why not fix it?", when really these are cases where further action needs to be taken on Rigetti's side due to unforeseen causes. 

Maybe I'm misinterpreting the intent here, but I take this as "you probably can't code your way out of this problem, and we need to change something internally".

Reply:
Yea, I totally agree with the approach - just not the term. But, it's just a name after all 👍

Reply:
I could call this `Unexpected` as well to match the top one? Maybe that better conveys the intent of the error—we don't expect this to ever be returned, not "we return this on purpose when there's a known bug".

Reply:
> I could call this Unexpected as well to match the top one? Maybe that better conveys the intent of the error—we don't expect this to ever be returned, not "we return this on purpose when there's a known bug".

I did that

Reply:
Thanks, @dbanty!